Add support for showing diagnostics in the editor

vlaaad opened the pull request:
This is a first changeset in series of adding linting support to the editor. It adds diagnostics as a concept and renders them in the code view. Diagnostics are regions (aka cursor ranges) with some extra fields that will be published by language servers to the editor and shown in code view. This changeset allows showing diagnostics, but does not yet have any support for actually creating diagnostics. To test it, you can do the following:

1. Open code view and select some text (multi-selection works too)
2. Evaluate this in the REPL:

       (let [code-view (-> 0
                           (dynamo.graph/targets-of :build-settings)
                           ffirst
                           (dynamo.graph/targets-of :_node-id)
                           ffirst
                           (dynamo.graph/node-value :active-view))
             ranges (dynamo.graph/node-value code-view :cursor-ranges)
             resource (dynamo.graph/node-value (dynamo.graph/node-value code-view :resource-node) :resource)
             diagnostics (mapv #(let [s (rand-nth [:error :warning :information :hint])]
                                  (assoc % :type :diagnostic
                                           :severity s
                                           :message (str "Severity is " s)))
                               ranges)]
            (:status (dynamo.graph/transact
                       (dynamo.graph/update-property 0 :resource->diagnostics assoc resource diagnostics))))

I don't plan to merge this to `editor-dev` yet. Instead, I'll merge other LSP-related changesets to this branch and, when everything is done, I will finally merge it to `editor-dev`.

Related to #3885

vlaaad committed:
Add support for showing diagnostics in the editor

This is a first changeset in series of adding linting support to the editor. It adds diagnostics as a concept and renders them in the code view. Diagnostics are regions (aka cursor ranges) with some extra fields that will be published by language servers to the editor and shown in code view. This changeset allows showing diagnostics, but does not yet have any support for actually creating diagnostics. To test it, you can do the following:

1. Open code view and select some text (multi-selection works too)
2. Evaluate this in the REPL:

       (let [code-view (-> 0
                           (dynamo.graph/targets-of :build-settings)
                           ffirst
                           (dynamo.graph/targets-of :_node-id)
                           ffirst
                           (dynamo.graph/node-value :active-view))
             ranges (dynamo.graph/node-value code-view :cursor-ranges)
             resource (dynamo.graph/node-value (dynamo.graph/node-value code-view :resource-node) :resource)
             diagnostics (mapv #(let [s (rand-nth [:error :warning :information :hint])]
                                  (assoc % :type :diagnostic
                                           :severity s
                                           :message (str "Severity is " s)))
                               ranges)]
            (:status (dynamo.graph/transact
                       (dynamo.graph/update-property 0 :resource->diagnostics assoc resource diagnostics))))

I don't plan to merge this to `editor-dev` yet. Instead, I'll merge other LSP-related changesets to this branch and, when everything is done, I will finally merge it to `editor-dev`.

Related to #3885

matgis commented on editor/src/clj/editor/code/view.clj:
```diff
@@ -70,10 +73,10 @@
-  (assert (or (nil? fill) (instance? Paint fill)))
-  (assert (or (nil? stroke) (instance? Paint stroke)))
-  (assert (instance? CursorRange cursor-range))
+  {:pre [(case type (:range :underline :word :diagnostic) true false)
```
Let's keep this list sorted!

matgis commented on editor/src/clj/editor/code/view.clj:
```diff
@@ -251,7 +258,8 @@
                (.fillRect gc (.x r) (.y r) (.w r) (.h r)))
-      :underline nil)))
+      :underline nil
+      :diagnostic nil)))
```
Could this be something more generic? `:squiggle` or `:issue`, perhaps? The `CursorRangeDrawInfo` `type` is typically associated with the type of visual element, separate from the `CursorRange` `type`.

vlaaad replied:
Good point, I'll make it a `:squiggle`

matgis commented on editor/src/clj/editor/code/data.clj:
```diff
@@ -2426,7 +2426,8 @@
     (when-some [clickable-region (some (fn [region]
-                                         (when (and (some? (:on-click! region))
+                                         (when (and (or (some? (:on-click! region))
+                                                        (= :diagnostic (:type region)))
```
It feels a bit strange having a special case for `:diagnostic` here. Perhaps we should treat hover popups as a higher-level concept?

vlaaad replied:
I agree that we should disentangle diagnostics from hovers.

matgis commented on editor/src/clj/editor/code/view.clj:
```diff
@@ -2428,6 +2460,48 @@
+                  (g/with-auto-evaluation-context evaluation-context
+                    (let [hovered-element (g/node-value view-node :hovered-element evaluation-context)]
+                      (if (and (= :region (:type hovered-element))
+                               (= :diagnostic (:type (:region hovered-element))))
```
Perhaps instead of hard-coding it to `:diagnostic` regions, we could associate a `:hover-view` function similar to `:on-click!` that returns a cljfx map for us to display here?

vlaaad replied:
Hmm yes, I like the idea of decomplecting hovers from diagnostics. Will do!

matgis commented on editor/src/clj/editor/fxui.clj:
```diff
@@ -65,6 +66,22 @@
+(defn tuple-lifecycle [& lifecycles]
+  (let [lifecycles (vec lifecycles)
+        len (count lifecycles)
+        instance-meta {`fx.component/instance #(mapv fx.component/instance %)}]
```
We could move this out and reuse it for all calls to `tuple-lifecycle`, perhaps?

vlaaad replied:
Yes, good point!

matgis commented on editor/src/clj/editor/fxui.clj:
```diff
@@ -451,6 +468,40 @@
+                                   {:on [{:fx/type fx/ext-get-ref :ref ::on} anchor-x anchor-y]})
+                          :desc (-> props
+                                    (dissoc :desc :showing :anchor-x :anchor-y)
+                                    (assoc :fx/type :popup))}}
```
Should we use `cljfx.fx.popup/lifecycle` here instead of the `:popup` keyword?

vlaaad replied:
Ah good catch, forgot about that!

matgis commented on editor/src/reveal/editor/reveal.clj:
```diff
@@ -2,12 +2,14 @@
   (:require [clojure.main :as m]
             [clojure.string :as str]
             [dynamo.graph :as g]
+            editor.code.data
```
Wrap in `[]`? Although, do we even need to require it?

vlaaad replied:
Wrapping in `[]` is only needed if you want to do something with that ns, e.g. alias it. We only need to require it so imports can work — if we don't require the ns, `editor.code.data.*` classes won't exist and imports will throw exceptions.

vlaaad replied:
Btw, I know some Clojure linters like clj-kondo warn on unused `[]`-wrapped requires, but don't warn on.. uhm.. naked requires, because naked require signals that the ns is required for side effects, which is the case here.

matgis replied:
Ah, I see.

vlaaad committed:
Initial LSP manager implementation (#7138)

This changeset adds initial LSP manager implementation:
- starting and stopping language servers (currently only from the REPL);
- syncing in-memory code resource modifications to language servers (open/change/close);
- displaying language server diagnostics in the code editor view.

LSP support is structured in layers:
- base (implemented in `editor.lsp.base`) — base communication protocol, defines a protocol for sending and receiving strings;
- JSON-RPC (implemented in `editor.lsp.jsonrpc`) — bidirectional implementation of JSON-RPC protocol (both server and client can send requests and notifications to each other) on top of a base layer;
- server (implemented in `editor.lsp.server`) — LSP server lifecycle (initialize - serve requests - shutdown) on top of a JSON-RPC layer;
- manager (implemented in `editor.lsp`) — integrates the editor with a zoo of language servers.

Some peculiarities:
- LSP manager implements atomic removal of a language server by keeping track of channels input reads from. This list of channels (`:channels` in the manager state map) always includes the main input channel that is used by the editor to submit commands. When a language server is started, we add its output channel to the list of the manager's input channels. LSP manager process processes messages from this list as they come sequentially, and when we receive a command to stop the language server, we remove its output channel from the manager's inputs, which effectively prevents us from receiving any notification from it ever again, even if it's currently in the process of submitting new notification to the manager.
- We track open documents in terms of resources, not resource nodes, because resource sync can re-create nodes for the same resources;
- Code views might start with a resource node for some resource, then due to an external change get a different resource node with the same resource, then due to in-editor rename end up with a new resource node for another resource.
- For incremental updates, we make `splice-lines` to keep a ring of the last 16 changes. We need more than one because sometimes a single code editor data change (e.g. inserting line break) performs multiple `splice-lines` modifications (e.g. add a new line, then fix indentation). This also makes incremental updates work on redo after multiple modifications that merge into a single undo step.

Note: this is not ready to be released to our users, so it will be merged into the DEFEDIT-3885 branch.

Related to #3885

vlaaad committed:
Launch language servers from editor scripts (#7177)

Related to #3885

vlaaad committed:
Implement LSP file watchers (#7189)

While LSP does not define a static configuration for file watchers and recommends using dynamic registration instead, some language servers (like vscode-luacheck and lua-language-server) use static configuration defined using non-LSP vscode API. I mirrored the configuration by adding `:watched-files` server description key. The initial API only defines a `:pattern` glob to watch for, but later we might add other watcher options, i.e. excluding from the watcher some file change type (created/changed/deleted).

Notes:

- I added a `Globs` class that is based on `sun.nio.fs.Globs`, the main reason for forking it is to ensure the same behavior w.r.t. path separators on all platforms, because we use Unix-style path separators for resources on all platforms;
- I stopped hiding dot files since it's useful to view and edit them from the editor.

Related to #3885

vlaaad committed:
Start language servers in the project folder (#7236)

We now start language servers in the project folder so it's possible to specify launch command executable as a relative path.

Related to #3885

vlaaad committed:
Report defined globals in Lua configuration (#7276)

Report defined globals in Lua configuration

vlaaad commented:
@matgis This branch now can run the latest release of our fork of lua-language-server (https://github.com/defold/lua-language-server/releases/download/v0.0.3/release.zip). You also reviewed every individual commit to this branch. Should we merge it to `editor-dev`? Or perhaps you want to review it as a whole too?

vlaaad committed:
Correctly handle paths on win32 (#7284)

* Fix LSP path handling on Windows

Related to #3885

* Use Path::resolve to resolve the command path

matgis reviewed (approved):
LGTM!
